### PR TITLE
First draft for component support

### PR DESCRIPTION
**Pull request by TimLariviere**
Details here: https://github.com/fabulous-dev/FabPlaygroundCE/blob/main/FabPlaygroundCE/Playground.fs

_Simple stateful component_
```fs
module Counter =
    let body =
        view {
            let! count = state 0
            
            VStack() {
                Label($"Count is {count.Current}")
                    .centerHorizontal()
                
                Button'("Increment", fun () -> count.Set(count.Current + 1))
                Button'("Decrement", fun () -> count.Set(count.Current - 1))
            }
        }
```

https://github.com/fabulous-dev/Fabulous.MauiControls/assets/6429007/50e18c18-634f-4435-873b-752b63ee0135

_Parent-child communication_
```fs
module ParentChild =
    let child count =
        view {
            let! multiplier = state 1
            let countMultiplied = count * multiplier.Current
            
            VStack() {
                Label($"Count * {multiplier.Current} = {countMultiplied}")
                    .centerHorizontal()
                    
                Button'("Increment Multiplier", fun () -> multiplier.Set(multiplier.Current + 1))
                Button'("Decrement Multiplier", fun () -> multiplier.Set(multiplier.Current - 1))
            }
        }
        
    let parent =
        view {
            let! count = state 0
            
            VStack() {
                Label($"Count is {count.Current}")
                    .centerHorizontal()
                
                Button'("Increment Count", fun () -> count.Set(count.Current + 1))
                Button'("Decrement Count", fun () -> count.Set(count.Current - 1))
                    
                Component(child count.Current)
            }
        }
```

https://github.com/fabulous-dev/Fabulous.MauiControls/assets/6429007/b56c76fd-8174-4a04-90eb-7349151b8c45

### CODE REVIEW

File: samples/HelloWorld/App.fs [hunk 1]
```diff
@@ -1,83 +1,122 @@
...
+    let body =
+        view {
+            let! count = state 0
+            
+            VStack() {
+                Label($"Count is {count.Current}")
+                    .centerHorizontal()
+                
+                Button'("Increment", fun () -> count.Set(count.Current + 1))
+                Button'("Decrement", fun () -> count.Set(count.Current - 1))
```
Comment (edgarfgp) [inline samples/HelloWorld/App.fs:58]: Any thoughts on supporting SingleChildBuilder. Aka similar to SwiftUI

```fsharp
Button("Decrement") {
     count.Set(count.Current - 1)
}
```

Comment (TimLariviere) [reply samples/HelloWorld/App.fs:58]: Technically possible. Is it worth to break typical F# syntax for it though 🤔

Comment (TimLariviere) [reply samples/HelloWorld/App.fs:58]: In SwiftUI, it's using the trailing closure syntax which is part of the language itself and widely used.
Here, we would make an exception just for it in F#.

Comment (edgarfgp) [reply samples/HelloWorld/App.fs:58]: You are right. No worth faking `trailing closure syntax` just for this case

**Comment by TimLariviere** [comment]
Changed the ergonomics to be more in line with the current DSL:

```fs
type Fabulous.Maui.View with
    static member inline SimpleComponent() =
        Component() {
            Label("Hello Component")
                .centerHorizontal()
        }

    static member inline Counter() =
        Component() {
            let! count = state 0
            
            VStack() {
                Label($"Count is {count.Current}")
                    .centerHorizontal()
                
                Button'("Increment", fun () -> count.Set(count.Current + 1))
                Button'("Decrement", fun () -> count.Set(count.Current - 1))
            }
        }
        
    static member inline ParentChild_Child(count: int) =
        Component() {
            let! multiplier = state 1
            let countMultiplied = count * multiplier.Current
            
            VStack() {
                Label($"Count * {multiplier.Current} = {countMultiplied}")
                    .centerHorizontal()
                    
                Button'("Increment Multiplier", fun () -> multiplier.Set(multiplier.Current + 1))
                Button'("Decrement Multiplier", fun () -> multiplier.Set(multiplier.Current - 1))
            }
        }
    
    static member inline ParentChild_Parent() =
        Component() {
            let! count = state 0
            
            VStack() {
                Label($"Count is {count.Current}")
                    .centerHorizontal()
                
                Button'("Increment Count", fun () -> count.Set(count.Current + 1))
                Button'("Decrement Count", fun () -> count.Set(count.Current - 1))
                
                View.ParentChild_Child(count.Current)
            }
        }
        
    static member inline Child(count: BindingRequest<'T>) =
        Component() {
            let! boundCount = count
            
            VStack() {
                Label($"Child.Count is {boundCount.Current}")
                    .centerHorizontal()
                
                Button'("Increment", fun () -> boundCount.Set(boundCount.Current + 1))
                Button'("Decrement", fun () -> boundCount.Set(boundCount.Current - 1))
            }
        }
    
    static member inline BindingBetweenParentAndChild() =
        Component() {
            let! count = state 0
            
            VStack() {
                Label($"Parent.Count is {count.Current}")
                    .centerHorizontal()
                    
                Button'("Increment", fun () -> count.Set(count.Current + 1))
                Button'("Decrement", fun () -> count.Set(count.Current - 1))
                
                View.Child(ofState count)
            }
        }
    
    static member inline SharedContextBetweenComponents() =
        Component() {
            let sharedContext = ComponentContext()
            
            VStack() {
                View.Counter()
                    .withContext(sharedContext)
                    
                View.Counter()
                    .withContext(sharedContext)
            }
        }
```

**Comment by TimLariviere** [comment]
@edgarfgp btw I managed to remove the need for the `Border` intermediate control. Component acts as a proxy widget.
Might require some more tinkering but it's working :)

**Pull request merged**